hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,494 @@
+		}
+	}
+
+	return result
+}
+
+func (params ConversionParameters) copy() ConversionParameters {
+	result := params
+	result.ConversionContext = nil
+	result.ConversionContext = append(result.ConversionContext, params.ConversionContext...)
+
+	return result
+}
+
+func (params ConversionParameters) withAdditionalConversionContext(t Type) ConversionParameters {
+	result := params.copy()
+	result.ConversionContext = append(result.ConversionContext, t)
+
+	return result
+}
+
+type ConversionHandler func(builder *ConversionFunctionBuilder, params ConversionParameters) []dst.Stmt
+
+// TODO: There feels like overlap between this and the Storage Conversion Factories? Need further thinking to combine them?
+// ConversionFunctionBuilder is used to build a function converting between two similar types.
+// It has a set of built-in conversions and can be configured with additional conversions.
+type ConversionFunctionBuilder struct {
+	// TODO: Better way to let you fuss with this? How can you pick out what I've already put in here to overwrite it?
+	conversions []ConversionHandler
+
+	IdFactory             IdentifierFactory
+	CodeGenerationContext *CodeGenerationContext
+}
+
+// NewConversionFunctionBuilder creates a new ConversionFunctionBuilder with the default conversions already added.
+func NewConversionFunctionBuilder(idFactory IdentifierFactory, codeGenerationContext *CodeGenerationContext) *ConversionFunctionBuilder {
+	return &ConversionFunctionBuilder{
+		IdFactory:             idFactory,
+		CodeGenerationContext: codeGenerationContext,
+		conversions: []ConversionHandler{
+			// Complex wrapper types checked first
+			IdentityConvertComplexOptionalProperty,
+			IdentityConvertComplexArrayProperty,
+			IdentityConvertComplexMapProperty,
+
+			// TODO: a flip function of some kind would be kinda nice (for source vs dest)
+			IdentityAssignValidatedTypeDestination,
+			IdentityAssignValidatedTypeSource,
+			IdentityAssignPrimitiveType,
+			IdentityDeepCopyJSON,
+			IdentityAssignTypeName,
+		},
+	}
+}
+
+// AddConversionHandlers adds the specified conversion handlers to the end of the conversion list.
+func (builder *ConversionFunctionBuilder) AddConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(builder.conversions, conversionHandlers...)
+}
+
+// PrependConversionHandlers adds the specified conversion handlers to the beginning of the conversion list.
+func (builder *ConversionFunctionBuilder) PrependConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(conversionHandlers, builder.conversions...)
+}
+
+func (builder *ConversionFunctionBuilder) BuildConversion(params ConversionParameters) []dst.Stmt {
+	for _, conversion := range builder.conversions {
+		result := conversion(builder, params)
+		if len(result) > 0 {
+			return result
+		}
+	}
+
+	panic(fmt.Sprintf("don't know how to perform conversion for %T -> %T", params.SourceType, params.DestinationType))
```

Comment:
Suggest using the debug descriptions of the types to provide more information about each

Reply:
We don't have easy access to the full `types` collection here... we do have `builder.CodeGenerationContext.GetTypesInCurrentPackage()` but that's only the types in _this_ package so won't work for types in other packages.

I added `CodeGenerationContext.GetAllReachableTypes()` and used that instead.